netshare/learn/utils/word2vec_embedding.py

```diff
@@ -36,3 +41,90 @@ def word2vec_train(
     logger.info(f"Word2Vec model is saved at {model_path}")
 
     return model_path
+
+
+def get_word2vec_type_col(word2vec_cols: List[Any]) -> Dict[str, List[str]]:
+    dict_type_cols: Dict[str, List[str]] = {}
+    for col in word2vec_cols:
+        type = col.encoding.split("_")[1]
+        if type not in dict_type_cols:
+            dict_type_cols[type] = []
+        dict_type_cols[type].append(col.column)
+    return dict_type_cols
+
+
+def build_annoy_dictionary_word2vec(
+    df, model_path, word2vec_cols, word2vec_size, n_trees
+):
+
+    model = Word2Vec.load(model_path)
+    wv = model.wv
+
+    # type : [cols]
+    # ("ip": ["srcip", "dstip"])
+    # "port": ["srcport", "dstport"]
+    # "proto": ["proto"]
+    dict_type_cols = get_word2vec_type_col(word2vec_cols)
+    logger.info(dict_type_cols)
+
+    sets = []
+    dict_type_annDictPair = {}
+    for type, cols in dict_type_cols.items():
+        type_set = set(
+            list(itertools.chain.from_iterable([list(df[col]) for col in cols]))
+        )
+        type_ann = AnnoyIndex(word2vec_size, "angular")
+        type_dict = {}
+        index = 0
+
+        for ele in type_set:
+            type_ann.add_item(index, get_vector(model, str(ele), norm_option=True))
+            type_dict[index] = ele
+            index += 1
+        type_ann.build(n_trees)
+
+        dict_type_annDictPair[type] = (type_ann, type_dict)
```

Comment:
The type of this structure is pretty complex. It is pretty hard to understand later when you use `dict_type_annDictPair[word2vec_type_col][0],` what everything means, and this is a source of bugs.
WDYT about using `namedtuple` here?
Here's a suggestion:
```
from typing import NamedTuple

class annoyTypeDescription(NamedTuple):
   annoy_type: int
   annoy_dict: Dict[int, ???]  # I really don't know
```
and then your return value will be: `Dict[int, annoyTypeDescription]`

Reply:
I think so. I can change it.